File: mixin-padang-webface/src/main/js/padang/functions/dataset/PivotRows.ts
/*
 * Copyright (c) 2020-2023 Datamixin.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import FunctionPlan from "padang/plan/FunctionPlan";
import ParameterPlan from "padang/plan/ParameterPlan";
import ParameterPlanUtils from "padang/plan/ParameterPlanUtils";
import FunctionPlanRegistry from "padang/plan/FunctionPlanRegistry";

import ColumnFunction from "padang/functions/dataset/ColumnFunction";
import DatasetFunction from "padang/functions/dataset/DatasetFunction";

export default class PivotRows extends DatasetFunction {

	public static FUNCTION_NAME = "PivotRows";

	constructor(
		public dataset: string,
		public rows: string[],
		public columns: string,
		public values: string,
		public aggregate: string) {
		super(PivotRows.FUNCTION_NAME, dataset);
	}

	public static ROWS_PLAN: ParameterPlan = ParameterPlanUtils.createListPlan(
		"rows",
		"Rows",
		"Columns where values will be converted into row value",
		ColumnFunction.SPECIFIED_PLAN
	);
	public static COLUMNS_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
		"columns",
		"Columns",
		"Column where values will be converted into column names",
		"",
		ColumnFunction.getDefaultColumnKeysAssignable()
	);
	public static VALUES_PLAN: ParameterPlan = ParameterPlanUtils.createTextPlan(
		"values",
		"Values",
		"Column where values to be aggregated",
		"",
		ColumnFunction.getDefaultColumnKeysAssignable()
	);
	public static AGGREGATE_PLAN: ParameterPlan = ParameterPlanUtils.createLogicalPlan(
		"aggregate",
		"Aggregate",
		"Used aggregate function used to calculate value"
	);

	public static getPlan(): FunctionPlan {
		let plan = new FunctionPlan(
			PivotRows.FUNCTION_NAME,
			"Pivot Rows",
			"mdi-table-pivot",
			"Pivot rows and columns into multiple columns with aggregate"
		);
		let parameters = plan.getParameterList();
		parameters.add(DatasetFunction.DATASET_PLAN);
		parameters.add(PivotRows.ROWS_PLAN);
		parameters.add(PivotRows.COLUMNS_PLAN);
		parameters.add(PivotRows.VALUES_PLAN);
		parameters.add(PivotRows.AGGREGATE_PLAN);
		return plan;
	}

}

let registry = FunctionPlanRegistry.getInstance();
registry.registerPlan(PivotRows.getPlan());

File: mixin-padang-garut/src/main/python/garut/functions/dataset/pivot_rows.py
#
# Copyright (c) 2020-2023 Datamixin.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.#
from typing import Dict
from garut.functions.aggregate.sum import Sum
from garut.functions.aggregate.aggregate import Aggregate
from garut.parameter import Parameter
from garut.contexts.context import Context
from garut.instruction import InstructionFactory
from garut.functions.function import FunctionRegistry
from garut.functions.dataset.dataset_function import DatasetFunction


class PivotRows(DatasetFunction):

    FUNCTION_NAME = "PivotRows"

    ROWS = "rows"
    COLUMNS = "columns"
    VALUES = "values"
    AGGREGATE = "aggregate"
    NONE = "None"

    def execute(self, context: Context, options: Dict[str, any]) -> any:

        # Options
        dataset = self.getDataset(options)
        rows = options.get(PivotRows.ROWS, None)
        columns = options.get(PivotRows.COLUMNS, None)
        values = options.get(PivotRows.VALUES, None)

        # Pivot
        aggregate = options.get(PivotRows.AGGREGATE, Sum.FUNCTION_NAME)
        if aggregate != PivotRows.NONE:
            aggregateType = self._getType(aggregate)
            dataFrame = dataset.pivot_table(index=rows, columns=columns, values=values, aggfunc=aggregateType)
        else:
            dataFrame = dataset.pivot(index=rows, columns=columns, values=values)

        # Index
        if dataFrame.index.name in dataFrame.columns:
            dataFrame.index.name = "row_" + dataFrame.index.name
        dataFrame.reset_index(inplace=True)

        # Types
        return self.forceObjectToString(dataFrame)

    def _getType(self, aggregate: str):
        if aggregate == "":
            aggregate = Sum.FUNCTION_NAME
        module = registry.get(aggregate)
        type: Aggregate = module.type()
        return type.getFunction()


parameters = [
    DatasetFunction.DATASET,
    PivotRows.ROWS,
    PivotRows.COLUMNS,
    PivotRows.VALUES,
    Parameter(PivotRows.AGGREGATE, str, Sum.FUNCTION_NAME),
]
registry: FunctionRegistry = FunctionRegistry.getInstance()
registry.register(PivotRows.FUNCTION_NAME, PivotRows, parameters)

factory: InstructionFactory = InstructionFactory.getInstance()
factory.registerInstruction(PivotRows.FUNCTION_NAME)
